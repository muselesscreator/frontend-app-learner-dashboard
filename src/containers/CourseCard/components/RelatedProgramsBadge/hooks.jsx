import React from 'react';
import { useIntl } from '@edx/frontend-platform/i18n';

import { StrictDict } from 'utils';
import { selectors } from 'data/redux';
import { useCardValues } from 'hooks';

import messages from './messages';
import * as module from './hooks';

const { cardData } = selectors;

export const state = StrictDict({
  isOpen: (val) => React.useState(val), // eslint-disable-line
});

export const useRelatedProgramsBadgeData = ({ courseNumber }) => {
  const [isOpen, setIsOpen] = module.state.isOpen(false);
  const { formatMessage } = useIntl();
  const { numPrograms } = useCardValues(courseNumber, {
    numPrograms: cardData.numRelatedPrograms,
  });

  let programsMessage = '';
  if (numPrograms) {
    programsMessage = formatMessage(
      numPrograms === 1 ? messages.badgeLabelSingular : messages.badgeLabelPlural,
      { numPrograms },
    );
  }

  return {
    numPrograms,
    programsMessage,
    isOpen,
    openModal: React.useCallback(() => setIsOpen(true), [setIsOpen]),
    closeModal: React.useCallback(() => setIsOpen(false), [setIsOpen]),
  };
};

export default useRelatedProgramsBadgeData;
