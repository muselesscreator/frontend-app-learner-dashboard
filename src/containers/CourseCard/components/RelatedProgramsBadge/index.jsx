/* eslint-disable quotes */
import React from 'react';
import PropTypes from 'prop-types';

import { Button } from '@edx/paragon';
import { Program } from '@edx/paragon/icons';

import RelatedProgramsBadgeModal from 'containers/RelatedProgramsModal';
import useRelatedProgramsBadgeData from './hooks';

export const RelatedProgramsBadge = ({ courseNumber }) => {
  const {
    isOpen,
    openModal,
    closeModal,
    numPrograms,
    programsMessage,
  } = useRelatedProgramsBadgeData({ courseNumber });
  return (numPrograms > 0) && (
    <>
      <Button
        data-testid="RelatedProgramsBadge"
        variant="tertiary"
        size="sm"
        iconBefore={Program}
        onClick={openModal}
      >
        {programsMessage}
      </Button>
      <RelatedProgramsBadgeModal {...{ isOpen, closeModal, courseNumber }} />
    </>
  );
};
RelatedProgramsBadge.propTypes = {
  courseNumber: PropTypes.string.isRequired,
};

export default RelatedProgramsBadge;
