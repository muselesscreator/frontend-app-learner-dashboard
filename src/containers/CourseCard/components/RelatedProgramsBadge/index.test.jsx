import React from 'react';
import { shallow } from 'enzyme';

import useRelatedProgramsBadge from './hooks';
import RelatedProgramsBadge from '.';

jest.mock('containers/RelatedProgramsModal', () => 'RelatedProgramsModal');
jest.mock('./hooks', () => jest.fn());

const hookProps = {
  isOpen: true,
  openModal: jest.fn().mockName('useRelatedProgramsBadge.openModal'),
  closeModal: jest.fn().mockName('useRelatedProgramsBadge.closeModal'),
  numPrograms: 3,
  programsMessage: 'useRelatedProgramsBadge.programsMessage',
};

const courseNumber = 'test-course-number';

describe('RelatedProgramsBadge component', () => {
  test('empty render: no programs', () => {
    useRelatedProgramsBadge.mockReturnValueOnce({ ...hookProps, numPrograms: 0 });
    const el = shallow(<RelatedProgramsBadge courseNumber={courseNumber} />);
    expect(el.isEmptyRender()).toEqual(true);
  });
  test('snapshot: 3 programs', () => {
    useRelatedProgramsBadge.mockReturnValueOnce(hookProps);
    expect(shallow(<RelatedProgramsBadge courseNumber={courseNumber} />)).toMatchSnapshot();
  });
});
