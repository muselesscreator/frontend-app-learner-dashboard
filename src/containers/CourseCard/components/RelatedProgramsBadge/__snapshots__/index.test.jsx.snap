// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RelatedProgramsBadge component snapshot: 3 programs 1`] = `
<Fragment>
  <Button
    data-testid="RelatedProgramsBadge"
    onClick={[MockFunction useRelatedProgramsBadge.openModal]}
    size="sm"
    variant="tertiary"
  >
    useRelatedProgramsBadge.programsMessage
  </Button>
  <RelatedProgramsModal
    closeModal={[MockFunction useRelatedProgramsBadge.closeModal]}
    courseNumber="test-course-number"
    isOpen={true}
  />
</Fragment>
`;
